File: backend/app/models/order.py
import uuid
from datetime import datetime, timezone
from typing import List, Optional
from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4()), index=True
    )
    order_number: Mapped[str] = mapped_column(String(50), unique=True, index=True, nullable=False)
    user_id: Mapped[Optional[str]] = mapped_column(String(36), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    customer_name: Mapped[str] = mapped_column(String(150), nullable=False)
    customer_email: Mapped[str] = mapped_column(String(255), nullable=False)
    customer_phone: Mapped[str] = mapped_column(String(30), nullable=False)
    delivery_address: Mapped[str] = mapped_column(Text, nullable=False)
    delivery_city: Mapped[str] = mapped_column(String(100), default="Bengaluru", nullable=False)
    delivery_pincode: Mapped[str] = mapped_column(String(20), nullable=False)
    delivery_date: Mapped[str] = mapped_column(String(30), nullable=False)
    delivery_slot: Mapped[str] = mapped_column(String(50), nullable=False)
    instructions: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    
    subtotal: Mapped[float] = mapped_column(Float, nullable=False)
    tax_amount: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)
    delivery_fee: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)
    discount_amount: Mapped[float] = mapped_column(Float, default=0.0, nullable=False)
    total_amount: Mapped[float] = mapped_column(Float, nullable=False)
    
    status: Mapped[str] = mapped_column(String(30), default="CONFIRMED", nullable=False)  # PENDING, CONFIRMED, PREPARING, DECORATING, READY_FOR_DISPATCH, OUT_FOR_DELIVERY, DELIVERED, CANCELLED
    payment_status: Mapped[str] = mapped_column(String(30), default="SUCCESS", nullable=False)  # PENDING, SUCCESS, FAILED
    payment_method: Mapped[str] = mapped_column(String(50), default="RAZORPAY", nullable=False)
    payment_id: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False
    )

    user: Mapped[Optional["User"]] = relationship("User", back_populates="orders")
    items: Mapped[List["OrderItem"]] = relationship("OrderItem", back_populates="order", cascade="all, delete-orphan")


class OrderItem(Base):
    __tablename__ = "order_items"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4()), index=True
    )
    order_id: Mapped[str] = mapped_column(String(36), ForeignKey("orders.id", ondelete="CASCADE"), nullable=False)
    product_id: Mapped[str] = mapped_column(String(36), nullable=False)
    product_name: Mapped[str] = mapped_column(String(200), nullable=False)
    image_url: Mapped[str] = mapped_column(String(500), nullable=False)
    price: Mapped[float] = mapped_column(Float, nullable=False)
    quantity: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    flavour: Mapped[str] = mapped_column(String(100), nullable=False)
    weight: Mapped[str] = mapped_column(String(50), nullable=False)
    is_eggless: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    custom_message: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)

    order: Mapped["Order"] = relationship("Order", back_populates="items")

File: frontend/src/types/order.ts
export interface OrderItem {
  id?: string;
  product_id: string;
  product_name: string;
  image_url: string;
  price: number;
  quantity: number;
  flavour: string;
  weight: string;
  is_eggless: boolean;
  custom_message?: string;
}

export interface Order {
  id: string;
  order_number: string;
  customer_name: string;
  customer_email: string;
  customer_phone: string;
  delivery_address: string;
  delivery_city: string;
  delivery_pincode: string;
  delivery_date: string;
  delivery_slot: string;
  instructions?: string;
  subtotal: number;
  tax_amount: number;
  delivery_fee: number;
  discount_amount: number;
  total_amount: number;
  status: 'PENDING' | 'CONFIRMED' | 'PREPARING' | 'DECORATING' | 'READY_FOR_DISPATCH' | 'OUT_FOR_DELIVERY' | 'DELIVERED' | 'CANCELLED';
  payment_status: 'PENDING' | 'SUCCESS' | 'FAILED';
  payment_method: string;
  created_at: string;
  items: OrderItem[];
}

export interface KitchenQueueItem {
  id: string;
  order_number: string;
  customer_name: string;
  items_summary: string;
  delivery_date: string;
  delivery_slot: string;
  total_amount: number;
  status: string;
  is_urgent: boolean;
}

export interface AdminAnalyticsData {
  today_revenue: number;
  today_revenue_formatted: string;
  today_orders_count: number;
  pending_delivery_count: number;
  total_customers_count: number;
  active_baking_count: number;
  recent_orders: KitchenQueueItem[];
}

export interface CustomerSummary {
  id: string;
  full_name: string;
  email: string;
  phone?: string;
  role: string;
  orders_count: number;
  total_spent: number;
  created_at: string;
}
